ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+      if (prune_transactions(ps, &tangle, &sap) != RC_OK) {
+        goto cleanup;
+      }
+      cond_handle_timedwait(&ps->cond_pruning_service, &ps->lock_handle, PRUNING_RESCAN_INTERVAL_SHORT_MS);
+    }
+    lock_handle_lock(&ps->lock_handle);
+    end_timestamp = current_timestamp_ms();
+    log_info(logger_id, "Pruning from % " PRIu64 " to % " PRIu64 " took % " PRIu64 " milliseconds\n", start_index,
+             ps->last_pruned_snapshot_index, end_timestamp - start_timestamp);
+    cond_handle_timedwait(&ps->cond_pruning_service, &ps->lock_handle, PRUNING_RESCAN_INTERVAL_MS);
+  }
+
+cleanup:
+
+  if (iota_tangle_destroy(&tangle) != RC_OK) {
+    log_critical(logger_id, "Destroying tangle connection failed\n");
+  }
+
+  if (iota_spent_addresses_provider_destroy(&sap) != RC_OK) {
+    log_critical(logger_id, "Destroying spent addresses provider failed\n");
+  }
+
+  return NULL;
+}
+
+static retcode_t collect_transactions_to_prune(pruning_service_t *const ps, tangle_t const *const tangle,
+                                               spent_addresses_provider_t *const sap,
+                                               flex_trit_t const *const entry_point_hash,
+                                               hash243_set_t *const transactions_to_prune) {
+  retcode_t err;
+  DECLARE_PACK_SINGLE_TX(tx, txp, tx_pack);
+  collect_transactions_for_pruning_do_func_params_t params = {.tangle = tangle,
+                                                              .transactions_to_prune = transactions_to_prune};
+  bool entry_point_has_no_solid_entry_points_in_past_cone;
+  hash243_set_entry_t *iter = NULL, *tmp = NULL;
+  bool spent;
+
+  params.current_snapshot_index = ps->last_pruned_snapshot_index + 1;
+
+  ERR_BIND_GOTO(tangle_traversal_dfs_to_past(params.tangle, collect_transactions_for_pruning_do_func, entry_point_hash,
+                                             ps->conf->genesis_hash, NULL, &params),
+                err, cleanup);
+
+  entry_point_has_no_solid_entry_points_in_past_cone = true;
+  iter = NULL;
+  tmp = NULL;
+
+  HASH_ITER(hh, *params.transactions_to_prune, iter, tmp) {
+    if (iota_snapshot_has_solid_entry_point(ps->new_snapshot, iter->hash)) {
+      entry_point_has_no_solid_entry_points_in_past_cone = false;
+      break;
+    }
+  }
+
+  if (entry_point_has_no_solid_entry_points_in_past_cone) {
+    HASH_ITER(hh, *params.transactions_to_prune, iter, tmp) {
+      hash_pack_reset(&tx_pack);
+      ERR_BIND_GOTO(
+          iota_tangle_transaction_load_partial(tangle, iter->hash, &tx_pack, PARTIAL_TX_MODEL_ESSENCE_METADATA), err,
+          cleanup);
+      ERR_BIND_GOTO(iota_spent_addresses_service_was_tx_spent_from(sap, tangle, &tx, iter->hash, &spent), err, cleanup);
+      ERR_BIND_GOTO(tips_cache_remove(ps->tips_cache, iter->hash), err, cleanup);
+    }
+  }
+
+  if (!entry_point_has_no_solid_entry_points_in_past_cone) {
+    hash243_set_free(params.transactions_to_prune);
```

Comment:
Shouldn't it always be freed ? And there is a potential double free with what follows.

Reply:
No, it should be freed only if there was an error or any of the transactions in the past cone is a SEP,
there is not double free because `entry_point_has_no_solid_entry_points_in_past_cone == false` implies `err == RC_OK`, but i will make it explicit